### PR TITLE
[Issue #4] Test Case: Writing to file, given file path

### PR DESCRIPTION
**Pull request by Kinnich**
Adds script to write given text to a new file.
Test that file is written correctly.

Also minor changes and clean up to main testing file

### CODE REVIEW

File: python/write_file.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import os
+
+if len(sys.argv) != 3:
+    print("Don't forget to enter the output file path and a string")
+
+outfile = sys.argv[1]
+text = sys.argv[2]
+
+if os.path.exists(outfile):
+   print("File already exists, clean this up to be sure we're writing/reading a new file")
```
Comment (Kinnich) [inline python/write_file.py]: I normally would raise errors here but not sure what happens to the errors in the docker container and if we'd ever see them, or everything would grind to a halt. Any thoughts as to if I should be raising errors, or if that is going to be annoying in other languages?

Comment (nicholaides) [reply python/write_file.py]: Errors will only come out as text written to stderr (or sometimes to stdout) or as the process' exit code being non-zero. So, if an error is raised, at worst we'd know because the test fails and then we'd dig in to figure out what's going on.

I think ultimately, it might be nice to set up some sort of hook in pytest that, when a test fails, it prints all the stdout, stderr, exit code from the container. That would likely save a bunch of time in figuring out why a test is failing.

In this case specifically, if we want to ensure the test environment is set up right (i.e. not having that file already there), that shouldn't be the job of this script. ~Either the Dockerfile should perform the sanity check, or, if we end up using `docker exec` to test for the existence of and contents of the file, we could also use it to check for the *non*-existence of the file before we run the script.~

File: python/test_python.py [hunk 2]
```diff
@@ -82,33 +82,40 @@ def test_read_file(self, docker_container):
...
         assert str(docker_container.logs(), 'UTF-8') == expected
 
-# to run in command line:
-# docker run -it --rm rosetta-python
+
+class TestWriteFile:
+    """Test that a script, given a path to a file, can write to that file"""
+    @pytest.mark.parametrize(
+        'docker_container',
+        ['python write_file.py output.txt "Bob Barker"'],
```
Comment (nicholaides) [inline python/test_python.py]: Let's not leave it up to the script (i.e. `write_file.py`) to report the file's contents because it doesn't prove that the script actually wrote to the file, only that the script upper-cased and printed out the input text.

Instead, we could do something like this:
`['python write_file.py output.txt "Bob Barker"; cat output.txt'],`

For clarity, it might even be worth formatting it like this:
```python
["""
  python write_file.py output.txt "Bob Barker"
  cat output.txt
"""],
```

File: python/write_file.py (hunk 1 above)
Comment (nicholaides) [reply python/write_file.py]: Actually, instead of `docker exec`, we can put it the docker run command, like this:

```python
@pytest.mark.parametrize(
    'docker_container',
    ["""
      if [ -f output.txt ]; then
        echo 'Error: output.txt already exists'
        exit 1
      fi

      python write_file.py output.txt "Bob Barker"
      cat output.txt
    """],
```

File: python/write_file.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+with open(outfile, 'w') as f:
+    f.write(text.upper())
+
+# Open the file that was just written and print what you see
+with open(outfile, 'r') as f:
+    print(f.readline())
+
+# Clean up the file
+if os.path.exists(outfile):
```
Comment (nicholaides) [inline python/write_file.py]: We don't need to clean up the file b/c when the container exits, it will all go away.
BUT, if you run the script it locally, it'd be nice for it to clean up after itself, but the test should handle that

**Review by nicholaides — CHANGES_REQUESTED** [review]

File: python/write_file.py (hunk 1 above)
Comment (Kinnich) [reply python/write_file.py]: I can't figure out how to pass in the bash lines to the docker command.
This works:
```
    @pytest.mark.parametrize(
        'docker_container',
        [['/bin/sh', '-c', 'python write_file.py output.txt "Bob Barker"; cat output.txt']],
        indirect=True,
    )
```
but not
```
    @pytest.mark.parametrize(
        'docker_container',
        ['python write_file.py output.txt "Bob Barker"; cat output.txt'],
        indirect=True,
    )
```
and I still can't grasp why.

File: python/test_python.py (hunk 2 above)
Comment (Kinnich) [reply python/test_python.py]: Good point. Makes sense to have a super simple script

File: python/write_file.py (hunk 3 above)
Comment (Kinnich) [reply python/write_file.py]: I also can't figure out how to have the test delete the output file. I get that we don't need to clean it up if running in docker since the container will exit, so then i guess the test doesn't need to do anything. It's just when running it manually that we get the output.txt file. So I guess we just have to manually delete it.

File: python/test_python.py [hunk 4]
```diff
@@ -82,33 +82,41 @@ def test_read_file(self, docker_container):
...
-        assert docker_container.logs() == b'Argument Number 1\n'
+        assert docker_container.logs() == b'argument number 1\n'
 
 
 class TestReadJsonFile:
+    """Test that a JSON file is read correctly"""
     @pytest.mark.parametrize(
         'docker_container',
-        ['python read_json.py < person-records.json'],
+        ['python read_json.py < person-records.json'], # < person-records.json isn't doing anything here since the filepath is in the script
```
Comment (Kinnich) [inline python/test_python.py]: @nicholaides I realized that this test is weird. Is it ok just have the filepath given in the `read_json.py` script itself? or does it need to be passed in as an argv? 
As the command it written, `< person-records.json` is just being ignored.

File: python/write_file.py (hunk 3 above)
Comment (nicholaides) [reply python/write_file.py]: > It's just when running it manually that we get the output.txt file.

Maybe we should always attempt to delete a local `output.txt` but only if it exists (or try to delete and swallow the error when it doesn't exist). So, if it was run locally, then there would be something to delete, but if run in Docker, then it would "pointlessly" try to delete the file.

Comment (nicholaides) [reply python/write_file.py]: ~~Also, the deleting of `output.txt` could go at the end of the test, but it'd probably even be better if it went in the beginning of the test to ensure that there is no `output.txt` file leftover from a previous run.~~ (not relevant if we put a uuid in the filename)

Additionally/alternatively, we could name the file differently every time (like by putting a uuid in the filename). That way, it could never get confused by the output file from a previous run.

Additionally, we could put the output file in a directory (e.g. `f"tmp/output-{uuid}.txt"`) and then if we fail to clean up the files for some reason, it's no big deal-- we can just delete all the files in that directory if they accumulate.

File: python/test_python.py (hunk 4 above)
Comment (nicholaides) [reply python/test_python.py]: ~~Assuming this test is for #6, this is good-- I specifically want it to be reading from stdin.~~ (edit: I misunderstood, correcting below)

Comment (nicholaides) [reply python/test_python.py]: Sorry, I misunderstood. `read_json.py` should be reading from stdin, similar to the `stdin.py` script-- it shouldn't know there's a file named `person-records.json`.

File: python/write_file.py (hunk 1 above)
Comment (nicholaides) [reply python/write_file.py]: Sorry, I confused myself and maybe you, too. Here's a comment where I point out where I got confused: https://github.com/Kinnich/rosetta-io/pull/11/files#r1298866911

Therefore, we'll have to use the `[['/bin/sh', '-c', ...]]` approach when we want to run shell commands (like using the `...; cat output.txt` trick or when we want to redirect a file to a script's stdin). We can simplify it as `[['sh', '-c', ...]]]`

The problem was that somewhere along the line, our string gets turned into an array in order to be helpful

e.g. this:
`['python write_file.py output.txt "Bob Barker"; cat output.txt'],`

gets turned into this:
`[['python', 'write_file.py', 'output.txt', 'Bob Barker;', 'cat', 'output.txt']],`

So, `cat` is now the 4th argument to the script, instead of another command

File: python/read_json.py [hunk 5]
```diff
@@ -1,9 +1,15 @@
+"""Read JSON file from stdin line by line"""
 import json
 
-file_path = "person-records.json"
+json_input = ""
 
-with open(file_path, "r") as file:
-    data = json.load(file)
+while True:
```
Comment (Kinnich) [inline python/read_json.py:6]: @nicholaides I prob should've put this in a different PR because it's for a different test, but is this how we want to handle testing reading the JSON file?

File: python/test_python.py (hunk 4 above)
Comment (Kinnich) [reply python/test_python.py]: I updated the script and test. Lemme know if it looks good

File: python/write_file.py (hunk 3 above)
Comment (nicholaides) [reply python/write_file.py]: In our converasation yesterday we decided to punt on cleaning up the file or naming it w/ a UUID until it's a problem.

**Review by nicholaides — APPROVED** [review]
